### PR TITLE
Turn Dart 2 mode by default.

### PR DESCRIPTION
**Pull request by mraleph**

### CODE REVIEW

File: packages/flutter_tools/bin/xcode_backend.sh [hunk 1]
```diff
@@ -101,6 +101,8 @@ BuildApp() {
   local preview_dart_2_flag=""
   if [[ -n "$PREVIEW_DART_2" ]]; then
     preview_dart_2_flag="--preview-dart-2"
+  else
+    preview_dart_2_flag="--no-preview-dart-2"
```
Comment (mkustermann) [inline packages/flutter_tools/bin/xcode_backend.sh]: Due to the commit message I thought that this CL switches `--preview-dart-2` on by-default. Now here the `else` branch is the default which **switches it explicitly off** by-default. Why is that?

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl [hunk 2]
```diff
@@ -6,6 +6,8 @@ class {{pluginDartClass}} {
   static const MethodChannel _channel =
       const MethodChannel('{{projectName}}');
 
-  static Future<String> get platformVersion =>
-      _channel.invokeMethod('getPlatformVersion');
+  static Future<String> get platformVersion async {
+    final String version = await _channel.invokeMethod('getPlatformVersion');
+    return version;
```
Comment (mkustermann) [inline packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl]: Why was this change necessary? Is there a bug? Is it related to https://github.com/dart-lang/sdk/issues/31743 ?

File: packages/flutter_tools/lib/src/compile.dart [hunk 3]
```diff
@@ -124,10 +128,11 @@ Future<String> compile(
 /// restarts the Flutter app.
 class ResidentCompiler {
   ResidentCompiler(this._sdkRoot, {bool trackWidgetCreation: false,
-      String packagesPath})
+      String packagesPath, CompilerMessageConsumer compilerMessageConsumer: printError})
     : assert(_sdkRoot != null),
       _trackWidgetCreation = trackWidgetCreation,
-      _packagesPath = packagesPath {
+      _packagesPath = packagesPath,
+      stdoutHandler = new _StdoutHandler(consumer: compilerMessageConsumer) {
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/compile.dart]: Would it make sense to make the same change above on line 111 ?

File: packages/flutter_tools/lib/src/compile.dart [hunk 4]
```diff
@@ -12,8 +12,12 @@ import 'base/io.dart';
 import 'base/process_manager.dart';
 import 'globals.dart';
 
+typedef void CompilerMessageConsumer(String message);
+
 class _StdoutHandler {
-  _StdoutHandler() {
+  final CompilerMessageConsumer consumer;
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/compile.dart]: s/consumer/errorHandler/ ?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 5]
```diff
@@ -108,6 +108,29 @@ class _Compiler {
...
+
+    ResidentCompiler createCompiler() {
+      return new ResidentCompiler(
+          artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath),
+          packagesPath: PackageMap.globalPackagesPath,
+          trackWidgetCreation: trackWidgetCreation,
+          compilerMessageConsumer: reportCompilerMessage);
+    }
+
     compilerController.stream.listen((_CompilationRequest request) async {
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: You could re-write this using 

  try {
    await for (...) {}
  } finally {
    outputDillDirectory.deleteSync(recursive: true);
  }

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 6]
```diff
@@ -118,28 +141,30 @@ class _Compiler {
         while (compilationQueue.isNotEmpty) {
           final _CompilationRequest request = compilationQueue.first;
           printTrace('Compiling ${request.path}');
+          compiler ??= createCompiler();
+          suppressOutput = false;
           final String outputPath = await compiler.recompile(request.path,
             <String>[request.path],
             outputPath: outputDill.path,
           );
           // Copy output dill next to the source file.
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Move comment down to else branch?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 7]
```diff
@@ -118,28 +141,30 @@ class _Compiler {
...
-          final File kernelReadyToRun = await fs.file(outputPath).copy(
+          if (outputPath == null) {
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: If you know in which cases we have `outputPath == null`, please add a comment here explaining this. It seems weird to request a compilation without using the result!

**Review by mkustermann — APPROVED** [review]

File: packages/flutter_tools/lib/src/commands/build_aot.dart [hunk 8]
```diff
@@ -39,7 +39,10 @@ class BuildAotCommand extends BuildSubCommand {
       )
       ..addFlag('interpreter')
       ..addFlag('quiet', defaultsTo: false)
-      ..addFlag('preview-dart-2', negatable: false, hide: !verboseHelp)
+      ..addFlag('preview-dart-2',
+        defaultsTo: true,
+        negatable: true,
+	hide: !verboseHelp)
```
Comment (a14n) [inline packages/flutter_tools/lib/src/commands/build_aot.dart]: nit: indentation

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 1 above)
Comment (mraleph) [reply packages/flutter_tools/bin/xcode_backend.sh]: The option is passed by flutter tools from the outside. This shell script is never used on its own.

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl (hunk 2 above)
Comment (mraleph) [reply packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl]: `invokeMethod` returns `Future<dynamic>` not `Future<String>`.

File: packages/flutter_tools/lib/src/compile.dart (hunk 3 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/compile.dart]: Maybe in the future.

File: packages/flutter_tools/lib/src/compile.dart (hunk 4 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/compile.dart]: It consumes all output (stdout) which might not necessarily errors.

I think in the future we would want to stream compilation messages in a more machine readable format.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 5 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: We will refactor this class more in subsequent CLs.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 6 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 7 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done.

File: packages/flutter_tools/lib/src/commands/build_aot.dart (hunk 8 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/commands/build_aot.dart]: done.

**Comment by a-siva** [comment]
As part of this pull request we should also change dev/bots/test.dart to run the two sets of tests taht is flip the dart1 mode to run the tests with --no-preview-dart-2. Something like this
```
 const Map<String, ShardRunner> _kShards = const <String, ShardRunner>{
   'docs': _generateDocs,
   'analyze': _analyzeRepo,
-  'tests': _runTests,
+  'tests': _runTestsDart1,
+  'tests_dart2': _runTestsDart2,
   'coverage': _runCoverage,
 };
```
```
+Future<Null> _runTestsDart1() async {
+  if (Platform.isWindows) {
+    // AppVeyor platform is overloaded, won't be able to handle additional
+    // load of dart1 testing.
+    return;
+  }
+  _runTests(options: <String>['--no-preview-dart-2']);
+}
+
+Future<Null> _runTestsDart2() async {
+  _runTests(options: <String>['--preview-dart-2']);
+}
```

**Comment by mraleph** [comment]
I think we need to shard it, e.g. like we do now with test and test2.

Also notice that `flutter_tools` tests would just use Dart 2.

**Comment by mraleph** [comment]
@a-siva sorry I misunderstood what you were saying. I will apply the patch.

**Comment by devoncarew** [comment]
@a-siva, we should flip the default for `flutter analyze` as well, to keep this all consistent.

**Comment by a-siva** [comment]
That will have to wait until pub run can handle preview-dart-2

**Comment by devoncarew** [comment]
Why? I mean the default to `flutter analyzer --preview-dart-2`

File: packages/flutter_tools/lib/src/compile.dart [hunk 9]
```diff
@@ -233,4 +237,9 @@ class ResidentCompiler {
   void reset() {
     _server.stdin.writeln('reset');
   }
+
+  Future<dynamic> shutdown() {
+    _server.kill();
```
Comment (aam) [inline packages/flutter_tools/lib/src/compile.dart]: Do we have to kill the frontend server? Wouldn't sending 'quit' command work more cleanly?

Comment (mraleph) [reply packages/flutter_tools/lib/src/compile.dart]: Yeah, `quit` is probably cleaner, but I wanted to avoid handling the case when frontend server got into some weird state - handling that would require sending `quit`, waiting with timeout for server to exit, then killing it. 

We can clean this up after we land this CL.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 10]
```diff
@@ -118,28 +142,33 @@ class _Compiler {
...
           );
-          // Copy output dill next to the source file.
-          final File kernelReadyToRun = await fs.file(outputPath).copy(
-              request.path + '.dill');
-          compiler.accept();
-          compiler.reset();
-          request.result.complete(kernelReadyToRun.path);
+
+          // Check if the compiler produced the output. It could have crashed
+          // in which case outputPath would be null. In this case pass
```
Comment (aam) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: null is produced when compiler was not able to recover from compilation errors. If compiler crashed we wouldn't receive output at all as frontend server would not write <result> tag into stdout, I don't think.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 11]
```diff
@@ -118,28 +142,33 @@ class _Compiler {
...
+          // null upwards to the consumer and shutdown the compiler.
+          if (outputPath == null) {
+            request.result.complete(null);
+            await shutdown();
```
Comment (aam) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: It's not clear why we are destroying compiler if there were compiler was not able to compile input source, was not able to produce kernel file.

Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: To avoid bizarre errors because compiler got into a weird state.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 12]
```diff
@@ -108,6 +108,30 @@ class _Compiler {
         .createTempSync('output_dill');
     final File outputDill = outputDillDirectory.childFile('output.dill');
 
+    bool suppressOutput = false;
+    void reportCompilerMessage(String message) {
+      if (suppressOutput)
+        return;
+
+      if (message.startsWith('compiler message: Error: Could not resolve the package \'test\'')) {
```
Comment (aam) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: What is the reason for this logic and supporting infrastructure(like suppressOutput, reportCompilerMessage, compilerMessageConsumer)? Why do we handle this error specially and suppress the rest of compiler output?

Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: There is a test that tests that a nice message is produced when somebody tries to do `flutter test` on a wrong file (search for 'Failed to load test harness' in the whole file to find that place).

The suppression I added just to make output. We can definitely improve this later even further.

File: packages/flutter_tools/lib/src/compile.dart (hunk 9 above)
Comment (aam) [reply packages/flutter_tools/lib/src/compile.dart]: Arguably, if frontend server can get into weird state so that it doesn't respond to 'quit' command, we would want to know about it now. After all this is test suite that should help us uncover bugs like that.
Cleaning it up later is fine though.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 12 above)
Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: > The suppression I added just to make output.

Not sure I understand. It looks like the only place we use suppressOutput(set it to true) is here when we encounter this error. With this special handling we won't need to introduce suppressOutput flag, reportCompilerMessage function, compilerMessageConsumer parameter. Am I reading it wrong?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 11 above)
Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Similar to handling of `quit` command, I would argue that if compiler gets into weird state in presence of compilation errors, that is valuable to know and should be filed as an error against compiler.

File: packages/flutter/test/foundation/error_reporting_test.dart [hunk 13]
```diff
@@ -70,18 +66,18 @@ Future<Null> main() async {
       '^══╡ EXCEPTION CAUGHT BY ERROR HANDLING TEST ╞═══════════════════════════════════════════════════════\n'
       'The following assertion was thrown testing the error handling logic:\n'
       'Message goes here\\.\n'
-      '\'[^\']+flutter${dividerRegExp}test${dividerRegExp}foundation${dividerRegExp}error_reporting_test\\.dart\': Failed assertion: line [0-9]+ pos [0-9]+: \'false\'\n'
+      '\'[^\']+flutter/test/foundation/error_reporting_test\\.dart\': Failed assertion: line [0-9]+ pos [0-9]+: \'false\'\n'
```
Comment (Hixie) [inline packages/flutter/test/foundation/error_reporting_test.dart]: (following up from the other PR:)

Huh. Do we know how Windows developers feel about this? It seems surprising that we wouldn't use platform conventions in filenames in our stack traces.

cc @goderbauer

File: packages/flutter_tools/lib/src/commands/build_apk.dart [hunk 14]
```diff
@@ -15,10 +15,12 @@ class BuildApkCommand extends BuildSubCommand {
     usesPubOption();
 
     argParser
-      ..addFlag('preview-dart-2', negatable: false,  hide: !verboseHelp)
+      ..addFlag('preview-dart-2', defaultsTo: true, hide: !verboseHelp)
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: here and elsewhere: we should add help to these `preview-dart-2` flags

**Comment by Hixie** [comment]
LGTM

Please file bugs for the known performance regressions and mark them milestone 3. Thanks.

File: packages/flutter/test/foundation/error_reporting_test.dart (hunk 13 above)
Comment (mraleph) [reply packages/flutter/test/foundation/error_reporting_test.dart]: Note that when packages are involved Dart VM used to report errors like this `package:a\b\c` on Windows which is super confusing because `package` paths are supposed to be uris which use `/` not `\`.

File: packages/flutter_tools/lib/src/commands/build_apk.dart (hunk 14 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/commands/build_apk.dart]: Done.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 10 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Updated the comment.
